### PR TITLE
Feature/event patterns

### PR DESCRIPTION
**Pull request by Ap1234567**
depends on https://github.com/PythonFreeCourse/calendar/pull/126

### CODE REVIEW

File: app/templates/event/eventview.html [hunk 1]
```diff
@@ -29,6 +34,7 @@
                     </div>
                     <!--    Copy commented section to add another tab-->
                     <!--    <div class="tab-pane fade" id="(CHANGE_ME)" role="tabpanel" aria-labelledby="(CHANGE_ME)-tab">-->
+                    <!--    <div class="tab-pane fade" id="(CHANGE_ME)" role="tabpanel" aria-labelledby="(CHANGE_ME)-tab">-->
```
Comment (ivarshav) [inline app/templates/event/eventview.html:37]: Duplicate by mistake?

File: app/database/models.py [hunk 2]
```diff
@@ -50,6 +50,7 @@ class Event(Base):
 
     owner = relationship("User")
     owner_id = Column(Integer, ForeignKey("users.id"))
+    invitees = Column(String)
```
Comment (ivarshav) [inline app/database/models.py:53]: Please move this line up (after line 49).
Also, what is the diff between invitees  and participants?

File: app/internal/event.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+def get_invited_emails(invited_from_form):
+    invited_emails = []
+    for invited_email in invited_from_form.split(','):
+        invited_email = invited_email.strip()
+        try:
+            validate_email(invited_email, check_deliverability=False)
+            invited_emails.append(invited_email)
+        except EmailSyntaxError:
+            logging.debug(f'{invited_email} is not a valid email address')
```
Comment (ivarshav) [inline app/internal/event.py:28]: I remember you changed to error level :O

File: app/internal/event.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    return invited_emails
+
+
+def get_uninvited_regular_emails(session, owner_id, title, invited_emails):
+    regular_invitees = set()
+    invitees_query = session.query(Event).with_entities(Event.invitees)
+    similar_events_invitees = invitees_query.filter(Event.owner_id == owner_id,
+                                                    Event.title == title).all()
+    for record in similar_events_invitees:
+        for email in record[0].split(','):
```
Comment (yammesicka) [inline app/internal/event.py:39]: Maybe `regular_invitees.update`?

File: tests/client_fixture.py [hunk 5]
```diff
@@ -13,6 +13,18 @@ def client():
     return TestClient(app)
 
 
+@pytest.fixture(scope="session")
```
Comment (yammesicka) [inline tests/client_fixture.py:16]: Please use `session` from `conftest.py` instead :)

File: app/internal/event.py [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+            regular_invitees.add(email)
+
+    return regular_invitees.difference(set(invited_emails))
```
Comment (yammesicka) [inline app/internal/event.py:42]: Tip: `regular_invitees - set(invited_emails)` will work too

File: app/internal/event.py [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+        try:
+            validate_email(invited_email, check_deliverability=False)
+            invited_emails.append(invited_email)
+        except EmailSyntaxError:
+            logging.debug(f'{invited_email} is not a valid email address')
+
+    return invited_emails
+
+
+def get_uninvited_regular_emails(session, owner_id, title, invited_emails):
```
Comment (yammesicka) [inline app/internal/event.py:33]: Please add type annotations to your code

**Review by yammesicka — CHANGES_REQUESTED** [review]

File: app/database/models.py (hunk 2 above)
Comment (Ap1234567) [reply app/database/models.py:53]: It's just the email addresses of the ones invited, whether they are going to participate or not, or event if they exist as users

File: app/internal/event.py (hunk 3 above)
Comment (Ap1234567) [reply app/internal/event.py:28]: true, rebased

File: app/database/models.py [hunk 8]
```diff
@@ -47,6 +47,7 @@ class Event(Base):
     end = Column(DateTime, nullable=False)
```
Comment (yammesicka) [inline app/database/models.py:47]: See comments on the other CR

**Review by yammesicka — CHANGES_REQUESTED** [review]

File: tests/test_event.py [hunk 9]
```diff
@@ -28,6 +102,65 @@ def test_eventview_without_id(self, client):
         response = client.get("/event/view")
         assert response.status_code == HTTP_404_NOT_FOUND
 
+    def test_eventedit_missing_old_invites(self, client, user):
+        response = client.post("/event/edit",
+                               data=CORRECT_EVENT_FORM_DATA)
+        assert response.status_code == HTTP_302_FOUND
```
Comment (yammesicka) [inline tests/test_event.py:108]: Please use `.ok` instead.

File: tests/test_event.py [hunk 10]
```diff
@@ -28,6 +102,65 @@ def test_eventview_without_id(self, client):
...
+        response = client.post("/event/edit",
+                               data=WEEK_LATER_EVENT_FORM_DATA)
+        assert response.status_code == HTTP_302_FOUND
+        assert 'Same event happened 1 weeks before too. ' in \
+               response.headers['location'].replace('+', ' ')
+
+        response = client.post("/event/edit",
+                               data=TWO_WEEKS_LATER_EVENT_FORM_DATA)
+        assert response.status_code == HTTP_302_FOUND
+        assert 'Same event happened 1 weeks before too. ' in \
```
Comment (yammesicka) [inline tests/test_event.py:159]: Please use parenthesis instead using `\` (search in other places in the code too)

File: app/routers/event.py [hunk 11]
```diff
@@ -79,9 +132,12 @@ def update_event(event_id: int, event: Dict, db: Session
     return by_id(db=db, event_id=event_id)
 
 
-def create_event(db, title, start, end, owner_id, content=None, location=None):
+def create_event(db, title, start, end, owner_id, content=None, location=None,
+                 invitees=None):
     """Creates an event and an association."""
 
+    invitees_concatenated = ','.join(invitees or [])
```
Comment (yammesicka) [inline app/routers/event.py:139]: Great!

File: app/database/models.py [hunk 12]
```diff
@@ -50,6 +50,7 @@ class Event(Base):
...
```
Comment (yammesicka) [inline app/database/models.py:50]: Great job on this PR!

**Review by yammesicka — CHANGES_REQUESTED** [review]

File: app/templates/event/partials/edit_event_details_tab.html [hunk 13]
```diff
@@ -21,6 +21,11 @@
     <input type="text" name="location" placeholder="VC URL/Location">
 </div>
 
+<div class="form_row">
+    <label>Invited emails: </label>
```
Comment (yammesicka) [inline app/templates/event/partials/edit_event_details_tab.html:25]: Use `for` to bind the label to the input field

File: app/internal/event.py [hunk 14]
```diff
@@ -1,12 +1,70 @@
...
+def get_uninvited_regular_emails(session: Session,
+                                 owner_id: int,
+                                 title: str,
+                                 invited_emails: List[str]):
+    regular_invitees = set()
+    invitees_query = session.query(Event).with_entities(Event.invitees)
+    similar_events_invitees = invitees_query.filter(Event.owner_id == owner_id,
+                                                    Event.title == title).all()
+    for record in similar_events_invitees:
+        regular_invitees.update(record[0].split(','))
```
Comment (yammesicka) [inline app/internal/event.py:44]: Prefer splitting only 1 times

**Review by yammesicka — COMMENTED** [review]
Great job, almost finished :)

File: app/internal/event.py [hunk 15]
```diff
@@ -1,12 +1,73 @@
...
+
+def get_invited_emails(invited_from_form: str):
+    invited_emails = []
+    for invited_email in invited_from_form.split(','):
+        invited_email = invited_email.strip()
+        try:
+            validate_email(invited_email, check_deliverability=False)
+            invited_emails.append(invited_email)
+        except EmailSyntaxError:
+            logging.error(f'{invited_email} is not a valid email address')
```
Comment (yammesicka) [inline app/internal/event.py:30]: Prefer `logging.exception`

File: app/internal/event.py [hunk 16]
```diff
@@ -1,12 +1,73 @@
...
                             detail="VC type with no valid zoom link")
+
+
+def get_invited_emails(invited_from_form: str):
+    invited_emails = []
+    for invited_email in invited_from_form.split(','):
+        invited_email = invited_email.strip()
+        try:
+            validate_email(invited_email, check_deliverability=False)
+            invited_emails.append(invited_email)
```
Comment (yammesicka) [inline app/internal/event.py:28]: Is this line might raise an exception? If not - please write it outside the try/except :)

File: app/internal/event.py [hunk 17]
```diff
@@ -1,12 +1,73 @@
...
+                                 owner_id: int,
+                                 title: str,
+                                 invited_emails: List[str]):
+    invitees_query = session.query(Event).with_entities(Event.invitees)
+    similar_events_invitees = invitees_query.filter(Event.owner_id == owner_id,
+                                                    Event.title == title).all()
+    all_regular_invitees_concatenated = ''
+    for record in similar_events_invitees:
+        if record:
+            all_regular_invitees_concatenated += ',' + record[0]
```
Comment (yammesicka) [inline app/internal/event.py:45]: You join it and then split it. Can you make it more efficient? :)

File: app/templates/event/eventview.html [hunk 18]
```diff
@@ -8,6 +8,11 @@
     <body>
         <div class = "event_view_wrapper">
                 <!-- Temporary nav layout based on bootstrap -->
+                <div class="messages">
+                    {% for message in messages %}
+                        <h2>{{ message }}</h2>
```
Comment (yammesicka) [inline app/templates/event/eventview.html:13]: Why h2?

**Review by yammesicka — COMMENTED** [review]
Great job, I've added few notes :)

File: app/internal/event.py (hunk 17 above)
Comment (Ap1234567) [reply app/internal/event.py:45]: I don't see how I can do that without changing the models

Comment (yammesicka) [reply app/internal/event.py:46]: Just add a variable called `regular_invitees` before the `for` loop, and use `regular_invitees.add(record[0])` each iteration

File: app/routers/event.py [hunk 19]
```diff
@@ -73,11 +71,19 @@ async def create_new_event(request: Request, session=Depends(get_db)):
 
     event = create_event(session, title, start, end, owner_id, content,
                          location, invited_emails, category_id=category_id)
-    message = ''
+    messages = []
```
Comment (yammesicka) [inline app/routers/event.py:74]: Split to another function

File: app/internal/event.py [hunk 20]
```diff
@@ -17,25 +19,56 @@ def raise_if_zoom_link_invalid(location):
                             detail="VC type with no valid zoom link")
 
 
-def get_invited_emails(invited_from_form):
+def get_invited_emails(invited_from_form: str):
```
Comment (yammesicka) [inline app/internal/event.py:22]: Please add the return type

File: app/internal/event.py [hunk 21]
```diff
@@ -17,25 +19,56 @@ def raise_if_zoom_link_invalid(location):
...
 
     return invited_emails
 
 
-def get_uninvited_regular_emails(session, owner_id, title, invited_emails):
-    regular_invitees = set()
+def get_uninvited_regular_emails(session: Session,
+                                 owner_id: int,
+                                 title: str,
+                                 invited_emails: List[str]):
```
Comment (yammesicka) [inline app/internal/event.py:39]: Please add the return type :)

**Review by yammesicka — CHANGES_REQUESTED** [review]

File: app/internal/event.py (hunk 17 above)
Comment (Ap1234567) [reply app/internal/event.py:46]: record[0] is a string of the format 'a@a.com,b@b.com'. I switched instead to ```regular_invitees.update(record[0].split(','))```

**Review by yammesicka — APPROVED** [review]

**Pull request merged**